### PR TITLE
Server tls-crypt-v2 support

### PR DESCRIPTION
**Pull request by reynir**
Implement tls-crypt-v2 in the server allowing configurations with both tls-crypt-v2 and either tls-auth or tls-crypt.

### CODE REVIEW

**Review by reynir — COMMENTED** [review]
This is a first draft of tls-crypt-v2 support. I am not very pleased with it. `new_connection` now takes the first packet as input to determine if we are using tls-crypt-v2 or either tls-auth or tls-crypt in setups that allow both. The packet then has to be passed to `handle` to do remaining parsing. This is in my opinion poor ergonomic and I plan to experiment with a different interface.

File: src/engine.ml [hunk 1]
```diff
@@ -324,34 +327,110 @@ let server server_config ~is_not_taken server_ts server_now server_rng =
     server_ip server_config,
     port )
 
-let new_connection server =
+let new_connection server data =
   let open Result.Syntax in
+  (* XXX(reynir): for now only TCP server *)
+  let protocol = `Tcp in
```
[the remark this answers is not in the document]
Comment (hannesm) [reply src/engine.ml:333]: fine with that. I'm not so convinced of the `XXX` comment -- it doesn't add much to the code. if we want to track that server is TCP only right now, I'd prefer to have a GitHub issue open (and tbh, I'd open it once there's need for server on UDP).

File: src/engine.ml [hunk 2]
```diff
@@ -324,34 +327,110 @@ let server server_config ~is_not_taken server_ts server_now server_rng =
     server_ip server_config,
     port )
 
-let new_connection server =
+let new_connection server data =
   let open Result.Syntax in
+  (* XXX(reynir): for now only TCP server *)
```
Comment (reynir) [inline src/engine.ml:332]:
```suggestion
```

**Review by hannesm — APPROVED** [review]
Great! :)

**Comment by hannesm** [comment]
I reviewed by avoiding the whitespace changes (adding `?w=1` to the url). This looks good to me.

File: src/engine.ml [hunk 3]
```diff
@@ -1918,6 +1892,101 @@ let incoming state control_crypto buf =
...
+                  last_received = current_ts;
+                  last_sent = current_ts;
+                },
+                control_crypto,
+                actual_packet )
+        | Ok (_not_hard_reset_client, _key, _received, _linger) ->
+            Error (`Msg "invalid initial packet"))
+    (* `No_transition *)
+  in
+  incoming t control_crypto packet
```
Comment (hannesm) [inline src/engine.ml:1988]: in incoming, there's:
```OCaml
              | Packet.Hard_reset_client_v3 ->
                  ignore_udp_error (Error (`No_transition (ch, op, received)))
```

which we should remove now, shouldn't we? and the serve should then also in `incoming_control_server` deal with reset_3...

File: src/engine.ml [hunk 4]
```diff
@@ -1918,6 +1892,101 @@ let incoming state control_crypto buf =
...
+            tls_auth_or_tls_crypt (`Msg "server only supports tls-crypt-v2")
+        | Ok (Packet.Hard_reset_client_v3, _key, received, linger) ->
+            assert (Cstruct.is_empty linger);
+            (* decode and unwrap wKc *)
+            let* actual_packet, wkc =
+              Tls_crypt.Wrapped_key.of_cstruct received
+            in
+            let* tls_crypt, _metadata =
+              Tls_crypt.Wrapped_key.unwrap ~key:wrapping_key wkc
+            in
```
Comment (hannesm) [inline src/engine.ml:1957]: you likely know better -- is the wkc encrypted as the other parts of the control packets? according to `handle_client`, it looks like it is encrypted in wrap_and_out calling encrypt_and_out

**Pull request merged**